File: vespajlib/src/main/java/com/yahoo/slime/BinaryDecoder.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.slime;


import static com.yahoo.slime.BinaryFormat.decode_double;
import static com.yahoo.slime.BinaryFormat.decode_meta;
import static com.yahoo.slime.BinaryFormat.decode_type;
import static com.yahoo.slime.BinaryFormat.decode_zigzag;

final class BinaryDecoder {

    BufferedInput in;

    private final SlimeInserter slimeInserter = new SlimeInserter(null);
    private final ArrayInserter arrayInserter = new ArrayInserter(null);
    private final ObjectSymbolInserter objectInserter = new ObjectSymbolInserter(null, 0);

    public BinaryDecoder() {}

    public Slime decode(byte[] bytes) {
        return decode(bytes, 0, bytes.length);
    }
    public Slime decode(byte[] bytes, int offset, int length) {
        Slime slime = new Slime();
        in = new BufferedInput(bytes, offset, length);
        decodeSymbolTable(in, slime.symbolTable());
        decodeValue(slimeInserter.adjust(slime));
        if (in.failed()) {
            slime.wrap("partial_result");
            slime.get().setData("offending_input", in.getOffending());
            slime.get().setString("error_message", in.getErrorMessage());
        }
        return slime;
    }

    long read_bytes_le(int bytes) {
        long value = 0;
        int shift = 0;
        for (int i = 0; i < bytes; ++i) {
            long b = in.getByte();
            value |= (b & 0xff) << shift;
            shift += 8;
        }
        return value;
    }

    long read_bytes_be(int bytes) {
        long value = 0;
        int shift = 56;
        for (int i = 0; i < bytes; ++i) {
            long b = in.getByte();
            value |= (b & 0xff) << shift;
            shift -= 8;
        }
        return value;
    }

    Cursor decodeNIX(Inserter inserter) {
        return inserter.insertNIX();
    }

    Cursor decodeBOOL(Inserter inserter, int meta) {
        return inserter.insertBOOL(meta != 0);
    }

    Cursor decodeLONG(Inserter inserter, int meta) {
        long encoded = read_bytes_le(meta);
        return inserter.insertLONG(decode_zigzag(encoded));
    }

    Cursor decodeDOUBLE(Inserter inserter, int meta) {
        long encoded = read_bytes_be(meta);
        return inserter.insertDOUBLE(decode_double(encoded));
    }

    Cursor decodeSTRING(Inserter inserter, int meta) {
        int size = in.read_size(meta);
        byte[] image = in.getBytes(size);
        return inserter.insertSTRING(image);
    }

    Cursor decodeDATA(Inserter inserter, int meta) {
        int size = in.read_size(meta);
        byte[] image = in.getBytes(size);
        return inserter.insertDATA(image);
    }

    Cursor decodeARRAY(Inserter inserter, int meta) {
        Cursor cursor = inserter.insertARRAY();
        int size = in.read_size(meta);
        for (int i = 0; i < size; ++i) {
            decodeValue(arrayInserter.adjust(cursor));
        }
        return cursor;
    }

    Cursor decodeOBJECT(Inserter inserter, int meta) {
        Cursor cursor = inserter.insertOBJECT();
        int size = in.read_size(meta);
        for (int i = 0; i < size; ++i) {
            int symbol = in.read_cmpr_int();
            decodeValue(objectInserter.adjust(cursor, symbol));
        }
        return cursor;
    }

    Cursor decodeValue(Inserter inserter, Type type, int meta) {
        switch (type) {
        case NIX:     return decodeNIX(inserter);
        case BOOL:    return decodeBOOL(inserter, meta);
        case LONG:    return decodeLONG(inserter, meta);
        case DOUBLE:  return decodeDOUBLE(inserter, meta);
        case STRING:  return decodeSTRING(inserter, meta);
        case DATA:    return decodeDATA(inserter, meta);
        case ARRAY:   return decodeARRAY(inserter, meta);
        case OBJECT:  return decodeOBJECT(inserter, meta);
        }
        assert false : "should not be reached";
        return null;
    }

    void decodeValue(Inserter inserter) {
        byte b = in.getByte();
        Cursor cursor = decodeValue(inserter, decode_type(b), decode_meta(b));
        if (!cursor.valid()) {
            in.fail("failed to decode value");
        }
    }

    static void decodeSymbolTable(BufferedInput input, SymbolTable names) {
        int numSymbols = input.read_cmpr_int();
        final byte[] backing = input.getBacking();
        for (int i = 0; i < numSymbols; ++i) {
            int size = input.read_cmpr_int();
            int offset = input.getPosition();
            input.skip(size);
            int symbol = names.insert(Utf8Codec.decode(backing, offset, size));
            if (symbol != i) {
                input.fail("duplicate symbols in symbol table");
                return;
            }
        }
    }
}

File: client/go/internal/slime/binary_decoder.go
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
// Author: arnej

package slime

type binaryDecoder struct {
	buf     []byte
	pos     int
	symbols []string
	failed  bool
	failure string
}

var emptyByteSlice = make([]byte, 0)

func newBinaryDecoder(input []byte) binaryDecoder {
	return binaryDecoder{
		buf:    input,
		pos:    0,
		failed: false,
	}
}

func (in *binaryDecoder) fail(reason string) {
	in.failed = true
	in.failure = reason
}

func (in *binaryDecoder) getByte() byte {
	var b byte = 0
	if !in.failed {
		if in.pos < len(in.buf) {
			b = in.buf[in.pos]
			in.pos++
		} else {
			in.fail("buffer underflow")
		}
	}
	return b
}

func (in *binaryDecoder) getBytes(sz int) []byte {
	if !in.failed {
		if in.pos+sz <= len(in.buf) {
			start := in.pos
			endp := start + sz
			in.pos = endp
			return in.buf[start:endp]
		}
		in.fail("buffer underflow")
	}
	return emptyByteSlice
}

func (in *binaryDecoder) read_bytes_le(sz uint32) uint64 {
	var value uint64 = 0
	shift := 0
	for i := uint32(0); i < sz; i++ {
		b := uint64(in.getByte())
		value = value | (b << shift)
		shift = shift + 8
	}
	return value
}

func (in *binaryDecoder) read_bytes_be(sz uint32) uint64 {
	var value uint64 = 0
	shift := 56
	for i := uint32(0); i < sz; i++ {
		b := uint64(in.getByte())
		value = value | (b << shift)
		shift = shift - 8
	}
	return value
}

func (in *binaryDecoder) read_cmpr_int() int {
	next := in.getByte()
	var value int64 = int64(next & 0x7f)
	shift := 0
	for (next & 0x80) != 0 {
		shift += 7
		next = in.getByte()
		value = value | ((int64(next) & 0x7f) << shift)
		if shift > 32 || value > 0x7fff_ffff {
			in.fail("compressed int overflow")
			value = 0
			next = 0
		}
	}
	return int(value)
}

func (in *binaryDecoder) read_size(meta uint32) int {
	if meta != 0 {
		return int(meta - 1)
	}
	return in.read_cmpr_int()
}

func (in *binaryDecoder) decodeSymbolTable() {
	numSymbols := in.read_cmpr_int()
	in.symbols = make([]string, numSymbols)
	for i := 0; i < numSymbols; i++ {
		sz := in.read_cmpr_int()
		buf := in.getBytes(sz)
		in.symbols[i] = string(buf)
	}
}

func (in *binaryDecoder) decodeValue() Cursor {
	b := in.getByte()
	ty := decode_type(b)
	meta := decode_meta(b)
	switch ty {
	case NIX:
		return ValidNix
	case BOOL:
		return in.decodeBOOL(meta)
	case LONG:
		return in.decodeLONG(meta)
	case DOUBLE:
		return in.decodeDOUBLE(meta)
	case STRING:
		return in.decodeSTRING(meta)
	case DATA:
		return in.decodeDATA(meta)
	case ARRAY:
		return in.decodeARRAY(meta)
	case OBJECT:
		return in.decodeOBJECT(meta)
	}
	panic("bad type")
	// return InvalidNix
}

func (in *binaryDecoder) decodeBOOL(meta uint32) Cursor {
	return newBoolValue(meta != 0)
}

func (in *binaryDecoder) decodeLONG(meta uint32) Cursor {
	var encoded uint64 = in.read_bytes_le(meta)
	return newLongValue(decode_zigzag(encoded))
}

func (in *binaryDecoder) decodeDOUBLE(meta uint32) Cursor {
	var encoded uint64 = in.read_bytes_be(meta)
	return newDoubleValue(decode_double(encoded))
}

func (in *binaryDecoder) decodeSTRING(meta uint32) Cursor {
	sz := in.read_size(meta)
	image := in.getBytes(sz)
	return newStringValue(string(image))
}

func (in *binaryDecoder) decodeDATA(meta uint32) Cursor {
	sz := in.read_size(meta)
	image := in.getBytes(sz)
	return newDataValue(image)
}

func (in *binaryDecoder) decodeARRAY(meta uint32) Cursor {
	res := newArrayValue()
	sz := in.read_size(meta)
	for i := 0; i < sz; i++ {
		elem := in.decodeValue()
		res.addElement(elem)
	}
	return res
}

func (in *binaryDecoder) decodeOBJECT(meta uint32) Cursor {
	res := newObjectValue()
	sz := in.read_size(meta)
	for i := 0; i < sz; i++ {
		symbol := in.read_cmpr_int()
		if symbol < len(in.symbols) {
			name := in.symbols[symbol]
			elem := in.decodeValue()
			res.addElement(name, elem)
		} else {
			in.fail("symbol id out of range decoding object")
			return res
		}
	}
	return res
}
